lua/ui/game/painting/PaintingCanvas.lua

```diff
@@ -0,0 +1,332 @@
+    ---@param worldview WorldView
+    __init = function(self, worldview)
+        Bitmap.__init(self, worldview)
+
+        self.Trash = TrashBag()
+        self.Painting = CreatePainting()
+        self.Adapter = self.Trash:Add(GetPaintingCanvasAdapter(self))
+
+        self.WorldView = worldview
+
+        -- register us to render
+        self.WorldView:RegisterRenderable(self, tostring(self))
+
+        self.InhibitionSet = {}
+    end,
+
+    ---@param self UIPaintingCanvas
+    ---@param worldview WorldView
+    __post_init = function(self, worldview)
+        LayoutFor(self)
+            :Fill(worldview)
+            :ResetWidth()
+            :ResetHeight()
+            :Over(worldview)
+            :NeedsFrameUpdate(true)
+
+            -- for details about this, see also `HandleWorldViewEvent`
+            :DisableHitTest(true)
+            :End()
+    end,
+
+    ---@param self UIPaintingCanvas
+    OnDestroy = function(self)
+        Bitmap.OnDestroy(self)
+        self.Trash:Destroy()
+
+        -- deregister us to render
+        self.WorldView:UnregisterRenderable(tostring(self))
+    end,
+
+    --- Checks if the painting feature is disabled as a whole.
+    IsEnabledByGameOptions = function(self)
+        local gameOption = GetOptions("painting")
+        if gameOption == 'on' then
+            return true
+        elseif gameOption == 'observing' and (IsObserver() or SessionIsReplay()) then
+            return true
+        end
+
+        return false
+    end,
+
+    --- Checks if there are any reasons to be disabled.
+    ---@param self UIPaintingCanvas
+    IsInhibited = function(self)
+        return not table.empty(self.InhibitionSet)
+    end,
+
+    --- Lifts a reason to be disabled. Function is idempotent. Calling it twice is the same as calling it once.
+    ---@param self UIPaintingCanvas
+    ---@param reason string
+    LiftInhibition = function(self, reason)
+        self.InhibitionSet[reason] = nil
+    end,
+
+    --- Adds a reason to be disabled. Function is idempotent. Calling it twice is the same as calling it once. All active paintings are destroyed.
+    ---@param self UIPaintingCanvas
+    ---@param reason string
+    AddInhibition = function(self, reason)
+        self.InhibitionSet[reason] = true
+        self:CancelBrush()
+    end,
+
+    ---------------------------------------------------------------------------
+    --#region User interactions with the painting canvas
+
+    ---@param self UIPaintingCanvas
+    OnFrame = function(self)
+        -- feature: inhibit painting when holding shift
+        if IsKeyDown('Shift') then
+            self:AddInhibition('shift')
+        else
+            self:LiftInhibition('shift')
+        end
+
+        -- limitation: only enable painting when no unit is selected. This limitation enables us 
+        -- to use a wide range of (mouse) buttons that would otherwise be used by engine functions
+        if table.getn(GetSelectedUnits()) == 0 then
```

Comment:
- GetSelectedUnits() returns `nil` with an empty selection (this needs to be documented)
- `table.getn` errors on nil, unlike `table.getsize` which returns 0
  - the `table.getn` comparison might as well be removed since the empty selection returns `nil`, unless you know of a way to get an empty table from `GetSelectedUnits()`
```suggestion
        local selection = GetSelectedUnits()
        if not selection or table.getn(GetSelectedUnits()) == 0 then
```